Package inclusion guide document (#464)
Azure/CloudShell: docs/add-package-into-cloudshell.md

# Add a package into Cloud Shell

One of the main features of Cloud Shell is that it serves many different packages to its users. Users don't have to go through the hassle of installing packages themselves when using Cloud Shell.

## An installation work around

> For all packages that does not require root permissions, you are free to install it yourself. This includes python, powershell, nodejs packages. Here's an example as to how you can install a package like ORAS yourself.

Reviewer: ```suggestion
For all packages that does not require root permissions, you are free to install it yourself. This includes, but not limited to, Python, Powershell, Nodejs packages.

Here's an example as to how you can install a package like ORAS yourself:
```

> For all packages that does not require root permissions, you are free to install it yourself. This includes python, powershell, nodejs packages. Here's an example as to how you can install a package like ORAS yourself.

Reviewer 2: Nit: "For all packages that _do_ not require...."

```
cd ~
VERSION="1.2.0"
curl -LO "https://github.com/oras-project/oras/releases/download/v${VERSION}/oras_${VERSION}_linux_amd64.tar.gz"
mkdir -p ~/oras-install/
mkdir -p ~/.local/bin/
tar -zxf oras_${VERSION}_*.tar.gz -C oras-install/
mv ~/oras-install/oras ~/.local/bin/oras
rm -rf oras_${VERSION}_*.tar.gz oras-install/
```
Reference: https://oras.land/docs/installation#linux

> If you mount a storage account and have the package stored within your HOME directory (or any subfolders from HOME), you can store it for future use. In the example above, since the package is stored within `~/.local/bin`, it will be available for you on the next time you use Cloud Shell.  

Reviewer: ```suggestion
For persistence of your packages across multiple Cloud Shell sessions, it is recommended you enable the persistent storage with the "storage account" feature of Cloud Shell. In addition, install these packages in any directory in `$HOME`, since the `$HOME` directory is stored in the storage account.

In our above example you can see that the package is installed in the `~/.local/bin` directory.
```

## Requesting the package

As per security & compliance requirements, we are striving to have all our packages installed from Azure Linux (a.k.a Mariner). Azure Linux is a compliant Microsoft Linux Operating System. If you want your package to be available in Cloud Shell, we expect to have it downloadable from Azure Linux OS. 

### Check if package is available in Azure Linux

> The below instructions is assuming Azure Linux 2.0. Please note that Azure linux 3.0 is coming out soon.

Reviewer: I think we should move this after the `docker` command.

```suggestion
> [!NOTE]
> The instructions assume Azure Linux 2.0 because CloudShell is currently based on Azure Linux 2.0
```

> The below instructions is assuming Azure Linux 2.0. Please note that Azure linux 3.0 is coming out soon.

Reviewer 2: I believe Azure Linux 3.0 is technically out, but we haven't switched to it yet.  I think we can say
"The below instructions are for Azure Linux 2.0, the base OS that Cloud Shell currently uses.
Note: In the future, Cloud Shell will require the package to be in Azure Linux 3.0."

> Please check if the package is available on here - 

Reviewer: ```suggestion
You can check if a package exists in Azure Linux from your browser by going to the [Azure Linux Package repository](https://packages.microsoft.com/cbl-mariner/2.0/prod/base/x86_64/Packages/).
```

> Please check if the package is available on here - 

Reviewer 2: ```suggestion
Please check if the package is available here - 
```

https://packages.microsoft.com/cbl-mariner/2.0/prod/base/x86_64/Packages/

> You can also check via entering the below command.

Reviewer: ```suggestion
Alternatively, you can use `docker` to check if a package exists in Azure Linux. Run the following command in your terminal:
```

```
docker run mcr.microsoft.com/cbl-mariner/base/core:2.0 bash -c "tdnf list | grep <package name>"
 ```

### Requesting Azure Linux for the package

> You can skip this if the package is already available in Azure Linux

Reviewer: ```suggestion
You can skip this step if a package is already available in Azure Linux.
```

> You can skip this if the package is already available in Azure Linux

Reviewer 2: ```suggestion
You can skip this if the package is already available in Azure Linux.
```

> To get the Azure Linux team to install the package, please open an issue via [Azure Linux GitHub](https://github.com/microsoft/azurelinux/issues).

Reviewer: ```suggestion
To get the Azure Linux team to install a package, please open an issue on [Azure Linux GitHub repository](https://github.com/microsoft/azurelinux/issues).
```

> To get the Azure Linux team to install the package, please open an issue via [Azure Linux GitHub](https://github.com/microsoft/azurelinux/issues).

Reviewer 2: This implies they will install it for you.  I think they'll work with the issuer to add the package.  How about:  
```suggestion
To start the process of adding your package to Azure Linux, please open an issue via [Azure Linux GitHub](https://github.com/microsoft/azurelinux/issues).
```

### Create a pull request on Azure Cloud Shell GitHub

> Create a pull request on [Cloud Shell GitHub](https://github.com/Azure/CloudShell) for adding the package in. Please add the package within [base.Dockerfile](https://github.com/Azure/CloudShell/blob/master/linux/base.Dockerfile). 

Reviewer: ```suggestion
Finally create a pull request on [Cloud Shell repository](https://github.com/Azure/CloudShell) to add the package. Please add the package in [base.Dockerfile](https://github.com/Azure/CloudShell/blob/master/linux/base.Dockerfile)'s package list as shown below: 
```

> Please add the package at the end of the long tdnf update + installation.

Reviewer: ```suggestion
```

> Please add the package at the end of the long tdnf update + installation.

Author: It's mentioned above that all packages should meet Cloud Shell's package inclusion principles.

For all packages that do not require root permissions, you are free to install it yourself. This includes, but not limited to, Python, Powershell, Nodejs packages.

Here's an example as to how you can install a package like ORAS yourself:

```
cd ~
VERSION="1.2.0"
curl -LO "https://github.com/oras-project/oras/releases/download/v${VERSION}/oras_${VERSION}_linux_amd64.tar.gz"
mkdir -p ~/oras-install/
mkdir -p ~/.local/bin/
tar -zxf oras_${VERSION}_*.tar.gz -C oras-install/
mv ~/oras-install/oras ~/.local/bin/oras
rm -rf oras_${VERSION}_*.tar.gz oras-install/
```
Reference: https://oras.land/docs/installation#linux

For persistence of your packages across multiple Cloud Shell sessions, it is recommended you enable the persistent storage with the "storage account" feature of Cloud Shell. In addition, install these packages in any directory in `$HOME`, since the `$HOME` directory is stored in the storage account.

In our above example you can see that the package is installed in the `~/.local/bin` directory. Since it is within a directory within `$HOME`, the ORAS package will persistent across Cloud Shell sessions (assuming you are using a storage mount).

## Requesting the package

> As per security & compliance requirements, we are striving to have all our packages installed from Azure Linux (a.k.a Mariner). Azure Linux is a compliant Microsoft Linux Operating System. If you want your package to be available in Cloud Shell, we expect to have it downloadable from Azure Linux OS. 

Author: Thanks for suggestion @rgardler-msft . What is the appropriate term here? Would "Azure Linux package repository" be appropriate?

> 

Reviewer 3: Add a link to docs/package-inclusion-guide.md which provides more detailed information about what packages might be included.
